demos/https/iot_demo_https_s3_download_async.c

```diff
@@ -0,0 +1,732 @@
+ *
+ * @return `EXIT_SUCCESS` if the demo completes successfully; `EXIT_FAILURE` otherwise.
+ */
+int RunHttpsAsyncDownloadDemo( bool awsIotMqttMode,
+                 const char * pIdentifier,
+                 void * pNetworkServerInfo,
+                 void * pNetworkCredentialInfo,
+                 const IotNetworkInterface_t * pNetworkInterface )
+{
+    IOT_FUNCTION_ENTRY( int, EXIT_SUCCESS );
+
+    /* Unused parameters. */
+    ( void )awsIotMqttMode;
+    ( void )pIdentifier;
+    ( void )pNetworkServerInfo;
+
+    /* HTTPS Client library return status. */
+    IotHttpsReturnCode_t httpsClientStatus;
+
+    /* Configurations for the HTTPS connection. */
+    IotHttpsConnectionInfo_t connConfig = { 0 };
+    /* Handle identifying the HTTPS connection. */
+    IotHttpsConnectionHandle_t connHandle = IOT_HTTPS_CONNECTION_HANDLE_INITIALIZER;
+    /* Asynchronous request specific configurations. */
+    IotHttpsAsyncInfo_t asyncInfo = { 0 };
+
+    /* The location of the path within string IOT_DEMO_HTTPS_PRESIGNED_GET_URL. */
+    const char *pPath = NULL;
+    /* The length of the path within string IOT_DEMO_HTTPS_PRESIGNED_GET_URL. */
+    size_t pathLen = 0;
+    /* The location of the address within string IOT_DEMO_HTTPS_PRESIGNED_GET_URL. */
+    const char *pAddress = NULL;
+    /* The length of the address within string IOT_DEMO_HTTPS_PRESIGNED_GET_URL. */
+    size_t addressLen = 0;
+
+    /* The number of bytes we want to request with in each range of the file bytes. */
+    uint32_t numReqBytes = 0;
+    /* curByte indicates which starting byte we want to download next. */
+    uint32_t curByte = 0;
+
+    /* Retrieve the path location and length from IOT_DEMO_HTTPS_PRESIGNED_GET_URL. */
+    httpsClientStatus = IotHttpsClient_GetUrlPath( IOT_DEMO_HTTPS_PRESIGNED_GET_URL, (size_t)strlen( IOT_DEMO_HTTPS_PRESIGNED_GET_URL ), &pPath, &pathLen);
+    if (httpsClientStatus != IOT_HTTPS_OK)
+    {
+        IotLogError("An error occurred in IotHttpsClient_GetUrlPath() with error code %d.", httpsClientStatus);
+        return EXIT_FAILURE;
+    }
+
+    /* Retrieve the address location and length from the IOT_DEMO_HTTPS_PRESIGNED_GET_URL. */
+    httpsClientStatus = IotHttpsClient_GetUrlAddress( IOT_DEMO_HTTPS_PRESIGNED_GET_URL, (size_t)strlen( IOT_DEMO_HTTPS_PRESIGNED_GET_URL ), &pAddress, &addressLen);
+    if (httpsClientStatus != IOT_HTTPS_OK)
+    {
+        IotLogError("An error occurred in IotHttpsClient_GetUrlAddress() with error code %d.", httpsClientStatus );
+        return EXIT_FAILURE;
+    }
+
+    /* Set the connection configurations. */
+    connConfig.pAddress = pAddress;
+    connConfig.addressLen = addressLen;
+    connConfig.port = IOT_DEMO_HTTPS_PORT;
+    /* We disable SNI here because the address specified includes the S3 bucket name. */
+    connConfig.flags |= IOT_HTTPS_DISABLE_SNI; /* Disable SNI, enable TLS, enable Persistent connections, is HTTP/1.1 */
```

Comment:
TLS is enabled by default and so are persistent connections. Those options can be changes with available flags. I will delete this comment since it is showing default behavior.